src/circuit_builder.rs

```diff
@@ -849,6 +851,56 @@ impl<C: HaloCurve> CircuitBuilder<C> {
         }
     }
 
+    /// Computes `[1 / n(s)] p`.
+    pub fn curve_mul_inv_endo<InnerC: HaloCurve<BaseField = C::ScalarField>>(
+        &mut self,
+        mul: CurveMulOp,
+    ) -> CurveMulEndoResult {
+        // We witness r = [1 / n(s)] p, then verify that [n(s)] r = p, and return r.
+        let CurveMulOp { scalar, point } = mul;
+
+        struct ResultGenerator<InnerC: HaloCurve> {
+            mul: CurveMulOp,
+            result: AffinePointTarget,
+            _phantom: PhantomData<InnerC>,
+        }
+
+        impl<InnerC: HaloCurve> WitnessGenerator<InnerC::BaseField> for ResultGenerator<InnerC> {
+            fn dependencies(&self) -> Vec<Target> {
+                vec![self.mul.scalar, self.mul.point.x, self.mul.point.y]
+            }
+
+            fn generate(
+                &self,
+                _constants: &Vec<Vec<InnerC::BaseField>>,
+                witness: &PartialWitness<InnerC::BaseField>,
+            ) -> PartialWitness<InnerC::BaseField> {
+                let scalar = witness.get_target(self.mul.scalar)
+                    .try_convert::<InnerC::ScalarField>().expect("Improbable");
+                let scalar_inv = scalar.multiplicative_inverse().expect("Can't invert zero");
+                let point = witness.get_point_target(self.mul.point);
+                let result = (InnerC::convert(scalar_inv) * point.to_projective()).to_affine();
+
+                let mut result_witness = PartialWitness::new();
+                result_witness.set_point_target(self.result, result);
+                result_witness
+            }
+        }
```

Comment:
Doesn't this compute `[1/s] p` instead of `[1/n(s)] p`?

Reply:
Oh yeah you're right. I fixed that but `curve_mul_inv_endo` is still failing; looking into it now.

Reply:
I think there's a problem with `curve_msm_endo`, so I think once that's fixed `curve_mul_inv_endo` should work.

Reply:
Agreed. It looks like it had the same two bugs that you fixed in the non-endo MSM method recently, so I added your fixes in there. It seems there another issue related to how we use the g^65 opening; looking into that now.